src/actions/requests.rs

```diff
@@ -203,45 +201,27 @@ impl RequestAction for Definition {
         let span = ctx.convert_pos_to_span(file_path.clone(), params.position);
         let analysis = ctx.analysis.clone();
 
-        // If configured start racer concurrently and fallback to racer result
-        let racer_receiver = {
-            if ctx.config.lock().unwrap().goto_def_racer_fallback {
-                Some(work_pool::receive_from_thread(
-                    move || {
-                        let cache = ctx.racer_cache();
-                        let session = ctx.racer_session(&cache);
-                        let location = pos_to_racer_location(params.position);
-
-                        racer::find_definition(file_path, location, &session)
-                            .and_then(|rm| location_from_racer_match(&rm))
-                    },
-                    WorkDescription("textDocument/definition-racer"),
-                ))
-            } else {
-                None
-            }
-        };
-
         match analysis.goto_def(&span) {
             Ok(out) => {
                 let result = vec![ls_util::rls_to_location(&out)];
                 trace!("goto_def (compiler): {:?}", result);
                 Ok(result)
             }
-            _ => match racer_receiver {
-                Some(receiver) => match receiver.recv() {
-                    Ok(Some(r)) => {
-                        trace!("goto_def (Racer): {:?}", r);
-                        Ok(vec![r])
-                    }
-                    Ok(None) => {
-                        trace!("goto_def (Racer): None");
-                        Ok(vec![])
-                    }
-                    _ => Self::fallback_response(),
-                },
-                _ => Self::fallback_response(),
-            },
+            // If analysis failed & `racer_completion` is enabled try racer
+            _ if ctx.config.lock().unwrap().racer_completion => {
```

Comment:
As in https://github.com/rust-lang/rls/pull/1150#discussion_r237099523, mind extracting that to a 
```rust
let racer_completion = {
    let config = ctx.config.lock().unwrap();
    config.racer_completion
};
```
above the `match`?
```

Reply:
The problem I have is that in the past the desugaring of the temporaries led to subtle issues, where we acquire a mutex but drop it later (unclear at which exact moment).
The statements in the inner scope were evaluated, where the lock should be uncontested but it turned out to be false - since std mutex isn't reentrant, it actually led to deadlocks.

In this concrete example we don't access the config mutex in the inner scope but it's not obvious - I had to look up every definition of the called `ctx.$func` - one could easily imagine that in this case it might access some configuration when creating a racer cache or it might be changed later to do so.

Could we then lock the config and branch on it in the `_` case? I know it's not as 'pure' and involves 1 more level of nesting but I rather err on the side of explicit and safe rather than silent errors that are hard to debug/track.

Reply:
I don't agree because it's adding extra lines and I don't see the benefit. Borrowing from mutex locks causes locks to continue perhaps longer than intended. However there is no borrow here, no `&` so rust's move semantics should have us covered (possible because it's a copy type).

That means if a later change needs to access the config it will not cause a deadlock. Worst case they just lock the mutex again, which isn't a big deal. But it'll be a fairly obvious refactor if that happens to lock once anyway.

See playground example I used to sanity check my understanding: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=835ea17582d8cb9e3b284c0611f519b9

Reply:
Your example, plus the fact we've talked about it so much probably shows I should change it and remove any doubt about how it works.